src/compute/src/compute_state/peek_result_iterator.rs

```diff
@@ -0,0 +1,294 @@
+            }
+        };
+
+        self.cursor.step_val(&self.storage);
+
+        Some(result)
+    }
+}
+
+impl<Tr, L> PeekResultIterator<Tr, L>
+where
+    for<'a> Tr: TraceReader<DiffGat<'a> = &'a Diff>,
+    for<'a> Tr::Key<'a>: ToDatumIter + IntoOwned<'a, Owned = Row> + Eq,
+    for<'a> Tr::Val<'a>: ToDatumIter,
+    for<'a> Tr::TimeGat<'a>: PartialOrder<mz_repr::Timestamp>,
+    L: Iterator<Item = Row>,
+{
+    /// Extracts and returns the row currently pointed at by our cursor. Returns
+    /// `Ok(None)` if our MapFilterProject evaluates to `None`. Also returns any
+    /// errors that arise from evaluating the MapFilterProject.
+    fn extract_current_row(&mut self) -> Result<Option<(Row, NonZeroI64)>, String> {
+        // TODO: This arena could be maintained and reused for longer,
+        // but it wasn't clear at what interval we should flush
+        // it to ensure we don't accidentally spike our memory use.
+        // This choice is conservative, and not the end of the world
+        // from a performance perspective.
+        let arena = RowArena::new();
+
+        let key_item = self.cursor.key(&self.storage);
+        let key = key_item.to_datum_iter();
+        let row_item = self.cursor.val(&self.storage);
+        let row = row_item.to_datum_iter();
+
+        let mut borrow = self.datum_vec.borrow();
+        borrow.extend(key);
+        borrow.extend(row);
+
+        if self.has_literal_constraints {
+            // The peek was created from an IndexedFilter join. We have to add those columns
+            // here that the join would add in a dataflow.
+            let datum_vec = borrow.deref_mut();
+            // unwrap is ok, because it could be None only if !has_literal_constraints or if
+            // the iteration is finished. In the latter case we already exited the while
+            // loop.
+            datum_vec.extend(self.current_literal.as_ref().unwrap().iter());
+        }
+        if let Some(result) = self
+            .map_filter_project
+            .evaluate_into(&mut borrow, &arena, &mut self.row_builder)
+            .map(|row| row.cloned())
+            .map_err_to_string_with_causes()?
+        {
+            let mut copies = Diff::ZERO;
+            self.cursor.map_times(&self.storage, |time, diff| {
+                if time.less_equal(&self.peek_timestamp) {
+                    copies += diff;
+                }
+            });
+            let copies: i64 = if copies.is_negative() {
+                return Err(format!(
+                    "Invalid data in source, saw retractions ({}) for row that does not exist: {:?}",
+                    -copies, &*borrow,
+                ));
+            } else {
+                copies.into_inner()
+            };
+            // if copies > 0 ... otherwise skip
+            if let Some(copies) = NonZeroI64::new(copies) {
+                Ok(Some((result, copies)))
+            } else {
+                Ok(None)
+            }
```

Comment:
The comment is somewhat misleading: We only test if `copies != 0` in the line below. We already tested that `copies` is not negative above, so it's technically not wrong, but... still confusing!

I would:
* Remove the comment, it doesn't add anything imo.
* Change the diff output type of the iterator to `NonZeroUsize`.

The second thing also safes an unwrap in the caller, which is nice.